tools/distpartitioning/globalids.py

```diff
@@ -0,0 +1,260 @@
+    alltoall_cpu( rank, world_size, recv_counts, send_counts)
+
+    # allocate buffers to receive global-ids
+    recv_shuffle_global_nids = []
+    for i in recv_counts:
+        recv_shufle_global_nids.append(torch.zeros([i.item()], dtype=torch.int64))
+
+    # Use node_data to lookup global id to send over.
+    send_nodes = []
+    for i in recv_nodes:
+        #list of node-ids to lookup
+        global_nids = i.tolist()
+        if (len(global_nids) != 0):
+            common, ind1, ind2 = np.intersect1d(node_data[:,3], global_nids, return_indices=True)
+            values = node_data[ind1,0]
+            send_nodes.append(torch.Tensor(values).type(dtype=torch.int64))
+        else:
+            send_nodes.append(torch.Tensor([]).type(dtype=torch.int64))
+
+    #send receive global-ids
+    alltoallv_cpu(rank, world_size, recv_shuffle_global_nids, send_nodes)
+
+    #form the lists with global-ids and orig-node-ids
+    recv_shuffle_global_ids = [x.tolist() for x in recv_global_ids]
+    shuffle_global_ids = list(itertools.chain(*recv_shuffle_global_ids))
+    global_nids = list(itertools.chain(*global_nids_ranks))
+
+    return shuffle_global_nids, global_nids 
+
+
+def get_shuffle_global_nids_edges(rank, world_size, edge_data, parts_map, node_data):
+    """Edges which are owned by this rank, may have global_nids whose shuffle_global_nids are NOT present locally.
+    This function retrieves shuffle_global_nids for such global_nids.
+
+    Parameters: 
+    -----------
+    rank : integer
+        rank of the process
+    world_size : integer
+        total no. of processes used
+    edge_data : numpy ndarray
+        edge_data (augmented) as read from the xxx_edges.txt file
+    parts_map : dictionary
+        global_nid to rank/partition-id mappins as determined by METIS
+    node_data : numpy ndarray
+        node_data (augmented) as read from xxx_nodes.txt file
+
+    Returns:
+    --------
+    numpy ndarray
+        edge_data, with two new columns (shuffle_global_src_id, shuffle_global_dst_id) which are global ids after data shuffling
+    """
+
+    #determine unique node-ids present locally
+    global_nids = np.unique(np.concatenate([edge_data[:, 0], edge_data[:, 1], node_data[:, 3]]))
+
+    #determine the rank which owns orig-node-id <-> partition/rank mappings
+    part_ids = np.array(list(map(list, parts_map.items())))
+    commons, ind1, ind2 = np.intersect1d(part_ids[:,0], global_nids, return_indices=True)
+    part_ids = part_ids[ ind1,:]
+
+    #form list of lists, each list includes global_nids whose mappings (shuffle_global_nids) needs to be retrieved.
+    #and rank will be the process which owns mappings of these global_nids
+    glbal_nids_ranks = []
+    for i in range(world_size):
+        if (i == rank):
+            glbal_nids_ranks.append([])
+            continue
+        not_owned_nodes = part_ids[:,0][part_ids[:,1] == i]
+        global_nids_ranks.append(not_owned_nodes)
```

Comment:
it seems there is a bug here. does the code really work?

Reply:
I apologize for the inconvenience. I was using a slightly updated version for my testing which is the working code based on this pull request. As a result, it will have all the fixes which are required for it to run correctly.